File: movies-api/src/main/java/com/ivanfranchin/moviesapi/movie/MoviesController.java
package com.ivanfranchin.moviesapi.movie;

import com.ivanfranchin.moviesapi.movie.dto.AddCommentRequest;
import com.ivanfranchin.moviesapi.movie.dto.CreateMovieRequest;
import com.ivanfranchin.moviesapi.movie.dto.MovieDto;
import com.ivanfranchin.moviesapi.movie.mapper.MovieDtoMapper;
import com.ivanfranchin.moviesapi.movie.model.Movie;
import com.ivanfranchin.moviesapi.userextra.dto.UpdateMovieRequest;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;
import java.time.Instant;
import java.util.List;

import static com.ivanfranchin.moviesapi.config.SwaggerConfig.BEARER_KEY_SECURITY_SCHEME;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/movies")
public class MoviesController {

    private final MovieService movieService;
    private final MovieDtoMapper movieMapper;

    @GetMapping
    public List<MovieDto> getMovies() {
        return movieService.getMovies().stream().map(movieMapper::toMovieDto).toList();
    }

    @GetMapping("/{imdbId}")
    public MovieDto getMovie(@PathVariable String imdbId) {
        Movie movie = movieService.validateAndGetMovie(imdbId);
        return movieMapper.toMovieDto(movie);
    }

    @Operation(security = {@SecurityRequirement(name = BEARER_KEY_SECURITY_SCHEME)})
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping
    public MovieDto createMovie(@Valid @RequestBody CreateMovieRequest createMovieRequest) {
        Movie movie = Movie.from(createMovieRequest);
        movie = movieService.saveMovie(movie);
        return movieMapper.toMovieDto(movie);
    }

    @Operation(security = {@SecurityRequirement(name = BEARER_KEY_SECURITY_SCHEME)})
    @PutMapping("/{imdbId}")
    public MovieDto updateMovie(@PathVariable String imdbId, @Valid @RequestBody UpdateMovieRequest updateMovieRequest) {
        Movie movie = movieService.validateAndGetMovie(imdbId);
        Movie.updateFrom(updateMovieRequest, movie);
        movie = movieService.saveMovie(movie);
        return movieMapper.toMovieDto(movie);
    }

    @Operation(security = {@SecurityRequirement(name = BEARER_KEY_SECURITY_SCHEME)})
    @DeleteMapping("/{imdbId}")
    public MovieDto deleteMovie(@PathVariable String imdbId) {
        Movie movie = movieService.validateAndGetMovie(imdbId);
        movieService.deleteMovie(movie);
        return movieMapper.toMovieDto(movie);
    }

    @Operation(security = {@SecurityRequirement(name = BEARER_KEY_SECURITY_SCHEME)})
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/{imdbId}/comments")
    public MovieDto addMovieComment(@PathVariable String imdbId,
                                    @Valid @RequestBody AddCommentRequest addCommentRequest,
                                    Principal principal) {
        Movie movie = movieService.validateAndGetMovie(imdbId);
        Movie.Comment comment = new Movie.Comment(principal.getName(), addCommentRequest.text(), Instant.now());
        movie.getComments().addFirst(comment);
        movie = movieService.saveMovie(movie);
        return movieMapper.toMovieDto(movie);
    }
}

File: movies-api/src/main/kotlin/com/ivanfranchin/moviesapi/movie/MoviesController.kt
package com.ivanfranchin.moviesapi.movie

import com.ivanfranchin.moviesapi.movie.dto.AddCommentRequest
import com.ivanfranchin.moviesapi.movie.dto.CreateMovieRequest
import com.ivanfranchin.moviesapi.movie.dto.MovieDto
import com.ivanfranchin.moviesapi.movie.mapper.MovieDtoMapper
import com.ivanfranchin.moviesapi.movie.model.Movie
import com.ivanfranchin.moviesapi.userextra.dto.MovieDTO
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import java.security.Principal
import java.time.Instant

@RestController
@RequestMapping("/api/movies")
class MoviesController(
    private val movieService: MovieService,
    private val movieMapper: MovieDtoMapper
) {

    @GetMapping
    fun getMovies(): List<MovieDto> {
        return movieService.getMovies().map { movieMapper.toMovieDto(it) }
    }

    @GetMapping("/{imdbId}")
    fun getMovie(@PathVariable imdbId: String): MovieDto {
        val movie = movieService.validateAndGetMovie(imdbId)
        return movieMapper.toMovieDto(movie)
    }

    @Operation(security = [SecurityRequirement(name = com.ivanfranchin.moviesapi.config.SwaggerConfig.BEARER_KEY_SECURITY_SCHEME)])
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping
    fun createMovie(@Valid @RequestBody createMovieRequest: CreateMovieRequest): MovieDto {
        val movie = Movie.from(createMovieRequest)
        val savedMovie = movieService.saveMovie(movie)
        return movieMapper.toMovieDto(savedMovie)
    }

    @Operation(security = [SecurityRequirement(name = com.ivanfranchin.moviesapi.config.SwaggerConfig.BEARER_KEY_SECURITY_SCHEME)])
    @PutMapping("/{imdbId}")
    fun updateMovie(
        @PathVariable imdbId: String,
        @Valid @RequestBody updateMovieRequest: MovieDTO.UpdateMovieRequest
    ): MovieDto {
        val movie = movieService.validateAndGetMovie(imdbId)
        Movie.updateFrom(updateMovieRequest, movie)
        val updatedMovie = movieService.saveMovie(movie)
        return movieMapper.toMovieDto(updatedMovie)
    }

    @Operation(security = [SecurityRequirement(name = com.ivanfranchin.moviesapi.config.SwaggerConfig.BEARER_KEY_SECURITY_SCHEME)])
    @DeleteMapping("/{imdbId}")
    fun deleteMovie(@PathVariable imdbId: String): MovieDto {
        val movie = movieService.validateAndGetMovie(imdbId)
        movieService.deleteMovie(movie)
        return movieMapper.toMovieDto(movie)
    }

    @Operation(security = [SecurityRequirement(name = com.ivanfranchin.moviesapi.config.SwaggerConfig.BEARER_KEY_SECURITY_SCHEME)])
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/{imdbId}/comments")
    fun addMovieComment(
        @PathVariable imdbId: String,
        @Valid @RequestBody addCommentRequest: AddCommentRequest,
        principal: Principal
    ): MovieDto {
        val movie = movieService.validateAndGetMovie(imdbId)
        val comment = Movie.Comment(
            username = principal.name,
            text = addCommentRequest.text,
            timestamp = Instant.now()
        )
        movie.comments.add(0, comment) // Kotlin's equivalent of addFirst
        val updatedMovie = movieService.saveMovie(movie)
        return movieMapper.toMovieDto(updatedMovie)
    }
}
